### ISSUE TITLE
Query param type safety only works for the first param.

### ISSUE THREAD

**Issue by Noahkoole** [drwpow/openapi-typescript#1525]
**Description**

When supplying multiple query params. it seems that when entering the first one correct it turns in to some sort of "Any" type. Allowing for any type of param and no autocomplete either.

<img width="247" alt="image" src="https://github.com/drwpow/openapi-typescript/assets/58473217/75970457-0d18-4f98-a57a-30072b47f525"></br>
First one autocompletes fine. </br>
<img width="233" alt="image" src="https://github.com/drwpow/openapi-typescript/assets/58473217/b6f2edef-107f-470e-a5c3-5f2f5eb8dc58"></br>
the second one does not allow for auto complete and does not throw a type error.</br>

<img width="329" alt="image" src="https://github.com/drwpow/openapi-typescript/assets/58473217/10205aa1-f7ae-4d0e-9fe2-99e391b8bfb3"></br>
this is the object that it expects

**Reproduction**

Have an endpoint that expects multiple query params and try to fill in more then 1 in the options payload.

**Expected result**

Type safety when typing multiple

**Checklist**

- [ ] I’m willing to open a PR (see [CONTRIBUTING.md](https://github.com/drwpow/openapi-typescript/blob/main/packages/openapi-fetch/CONTRIBUTING.md))

**Comment by Noahkoole** [comment]
With further investigating, this issue was introduced in the 0.9.0next. Which I downloaded for the middleware, will post under that thread

**Comment by drwpow** [comment]
This seems to exist in `0.8.x` as well. Will investigate a fix

**Comment by drwpow** [comment]
So I think this may have something to do with how TypeScript’s `extends` keyword works internally—it was originally designed for class inheritance, and thus it doesn’t prevent adding additional properties to the base, so long as the required keys are provided. So in your example, it will allow `loffeeeqfqnk` because it’s met its base requirements.

To be honest I’m not sure how to easily fix this—only allow the explicit `parameters` in the OpenAPI schema, and no more. I didn’t dig up any hints in the [official TS docs on conditional types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html), but perhaps some TS wizards reading this would be so kind as to investigate a fix? 🙏

**Comment by gittgott** [comment]
Hey @drwpow,

Just FYI, in TypeScript 5.4 there will be a new intrinsic type called [`NoInfer`](https://github.com/microsoft/TypeScript/pull/56794). I put together a minimal example of how that could potentially solve this problem. Here that is: [TypeScript Playground Link](https://www.typescriptlang.org/play?ts=5.4.0-dev.20240217#code/GYVwdgxgLglg9mABFApgZygRgDwBVEoAeqYAJmogN4AOAhgE60C2aAXFQI4gr0Ce7lADYwmMKAH52YEEwBGPAL5KAfAApUGdrgCUVAFCJE9FFBD0kGqAG49CvXoD0DxAEFBguAHcKDWWMZ8iADWKLwUUHCI8oh0aGgopIgwFpFcPLyItMCo9MgAFiiIAOYoYDwwEADkFBAIGIzJUEkUTCZ6lpiqNAzMbJzcfALColDsmAAMkwA0yOijiABEeTALiEoK2vZOiABycACSYMA8mfRw4IlQBcWl5RBGc-QVUOHXULzUhRGIAFYgGIhPHlaE0YBQ0J8IDBgDAEkkLNcSmUnlUanUoA0wFA9KBILAELMMAAmPAEYilchUOiMFgCNKDKjDMSSRDSOSKFTqObsPaHY70PDKXSUAwPUzmQnWWz2SxErrU3p0gb8RkiMRjSbjGaWdhLFZrJSbIA)

I'm not sure how soon TypeScript 5.4 could be required for openapi-fetch, but I believe that NoInfer would solve this issue once available.

**Comment by gittgott** [comment]
Hey @drwpow,

So it seems like there's two issues here.

One is that the params can have arbitrary keys outside of the specified ones. I found that TypeScript 5.4 will be adding a new intrinsic type called [`NoInfer`](https://github.com/microsoft/TypeScript/pull/56794) that might help with this. I put together a small example in the [TypeScript playground](https://www.typescriptlang.org/play?ts=5.4.0-beta#code/GYVwdgxgLglg9mABFApgZygRgDwBVEoAeqYAJmogN4AOAhgE60C2aAXFQI4gr0Ce7lADYwmMKAH52YEEwBGPADSIIcJk3BjeAfRilJiDPRhgA5gF8LAPgAUqDO1wBKKgChEieiigh6SO1ABuFzMXFwB6MMQAQUFBOAB3CgZZMUY+RABrFF4KKDhEeUQ6NDQUUkRjZHyuHl5EWmBUemQACxREExQwHhgIAHIKFTBDWmMoCoomLxd-TGsaBmY2Tm4+AWFRKHZMAAY9xAszR1CIxAA5OABJMGAeevo4cHKoNo6unogPdCgjaFzXqC8ajtPKIABWIAwiHiLVo4xgFDQwIgMGAMDKFT8r063V+A2UCBGYxcoEgsAQyG+ACY8ARiF1yFQ6IwWAIamsqBsxPppHJFAS1BpATo9OxDMZzFZbN92Bdrrd6HhLM5KG4vt5fJSMEEQjNqfNmUs2at+JyRGJtnsdkp-OwAEQtGB2g4WY5AA) to demonstrate this. Not 100% sure if it will be as easy as just wrapping the init parameter's type from ClientMethod in `NoInfer`, but that might work once TypeScript 5.4 is available and able to be used as a minimum version for this package.

The other issue is that once you add one optional parameter to the request, the rest of the optional parameters do not show up in autocomplete. This issue seems specific to version ^0.9.0 of openapi-fetch. I can look into that and make a PR if I find anything.

**Comment by clementAC** [comment]
I confirm I have both the problems 🥲

**Comment by SebKranz** [comment]
I wrapped the init parameter in `NoInfer` and it does seem to resolve the issue.
So a practical workaround for those on TS 4.5 would be to copy the type definitions for `Client` and `ClientMethod` into your own code and export the generated client 

```ts
 import { FetchResponse, MaybeOptionalInit, Middleware } from "openapi-fetch"
import {
  HasRequiredKeys,
  HttpMethod,
  MediaType,
  PathsWithMethod,
} from "openapi-typescript-helpers"

type ClientMethod<
  Paths extends Record<string, Record<HttpMethod, {}>>,
  Method extends HttpMethod,
  Media extends MediaType,
> = <
  Path extends PathsWithMethod<Paths, Method>,
  Init extends MaybeOptionalInit<Paths[Path], Method>,
>(
  url: Path,
  ...init: NoInfer<
    HasRequiredKeys<Init> extends never
      ? [(Init & { [key: string]: unknown })?] // note: the arbitrary [key: string]: addition MUST happen here after all the inference happens (otherwise TS can’t infer if it’s required or not)
      : [Init & { [key: string]: unknown }]
  >
) => Promise<FetchResponse<Paths[Path][Method], Init, Media>>

export interface Client<Paths extends {}, Media extends MediaType = MediaType> {
  /** Call a GET endpoint */
  GET: ClientMethod<Paths, "get", Media>
  /** Call a PUT endpoint */
  PUT: ClientMethod<Paths, "put", Media>
  /** Call a POST endpoint */
  POST: ClientMethod<Paths, "post", Media>
  /** Call a DELETE endpoint */
  DELETE: ClientMethod<Paths, "delete", Media>
  /** Call a OPTIONS endpoint */
  OPTIONS: ClientMethod<Paths, "options", Media>
  /** Call a HEAD endpoint */
  HEAD: ClientMethod<Paths, "head", Media>
  /** Call a PATCH endpoint */
  PATCH: ClientMethod<Paths, "patch", Media>
  /** Call a TRACE endpoint */
  TRACE: ClientMethod<Paths, "trace", Media>
  /** Register middleware */
  use(...middleware: Middleware[]): void
  /** Unregister middleware */
  eject(...middleware: Middleware[]): void
}

export api  = createClient<paths>(...) as Client<paths>
```

**Comment by drwpow** [comment]
Thanks all for the detailed info and debugging. This bug is top-of-mind for us, and does need to be fixed. Bugs like this are what’s still
keeping this library at 0.x rather than 1.0, especially tricky type bugs like this. Fortunately I think we’re at the end of the biggest ones, but have a few significant ones remaining like this.

PRs are welcome, of course! This is a major bug that does need addressing, as so many of you have pointed out. I personally can’t take this on for the next couple months, but will drop what I’m doing at any time to review PRs that address this.

But beyond a simple fix, one thing I’ve realized we’re missing currently is **more robust `*.test-d.ts` tests** that take full advantage of Vitest’s type utils. We have a few tests like this, yes, but I’m proposing taking all of the type checks out of the main test and almost completely separating runtime and type tests. Because we’re missing granular assertions that are leading to bugs like this as the TS “inference chain” gets other improvements.

This doesn’t have to happen in the same fix, it could be its own PR, but I think this change in testing would help fixing these a lot less daunting.

**Comment by ngraef** [comment]
I'm not able to reproduce this in `openapi-fetch@0.11.1`. If anyone is still seeing this issue (specifically, the "only the first query param is type checked" issue), could you please provide a more complete reproduction case including your tsconfig.json? I'd be interested to look into this.

**Comment by OliverJAsh** [comment]
Edit: this appears to be fixed now.

I noticed that "go to definition" doesn't work on params, and JSDoc comments do not appear.

I'm not sure if this is exactly the same issue as reported above, but it seems related. 

Here is a reduced test case:

```ts
import create from 'openapi-fetch';

/**
 * This was auto-generated by openapi-typescript.
 */
interface paths {
  readonly '/users/{username}': {
    readonly parameters: {
      readonly query?: never;
      readonly header?: never;
      readonly path?: never;
      readonly cookie?: never;
    };
    readonly get: {
      readonly parameters: {
        readonly query?: never;
        readonly header?: never;
        readonly path: {
          /**
           * Test JSDoc comment
           */
          readonly username: string;
        };
        readonly cookie?: never;
      };
      readonly requestBody?: never;
      readonly responses: {
        /** @description OK */
        readonly 200: {
          headers: {
            readonly [name: string]: unknown;
          };
          content: {
            readonly 'application/json': {
              readonly id: string;
            };
          };
        };
      };
    };
    readonly put?: never;
    readonly post?: never;
    readonly delete?: never;
    readonly options?: never;
    readonly head?: never;
    readonly patch?: never;
    readonly trace?: never;
  };
}

const api = create<paths>();

api.GET('/users/{username}', {
  params: {
    path: {
      // ❌ "Go to definition" doesn't work here
      // ❌ JSDoc comment doesn't appear in the tooltip
      username: 'naoufal',
    },
  },
});

```

Looking at the types, I suspect the issue is because `Init` is a type parameter with a constraint. I suspect this was made a type parameter just to deduplicate the type, but I don't think that's a correct usage of type parameters.

I would recommend fixing it like so:

```diff
 export type ClientMethod<
   Paths extends Record<string, Record<HttpMethod, {}>>,
   Method extends HttpMethod,
   Media extends MediaType,
-> = <Path extends PathsWithMethod<Paths, Method>, Init extends MaybeOptionalInit<Paths[Path], Method>>(
+> = <Path extends PathsWithMethod<Paths, Method>>(
   url: Path,
-  ...init: InitParam<Init>
-) => Promise<FetchResponse<Paths[Path][Method], Init, Media>>;
+  ...init: InitParam<MaybeOptionalInit<Paths[Path], Method>>
+) => Promise<FetchResponse<Paths[Path][Method], MaybeOptionalInit<Paths[Path], Method>, Media>>;

```

Whilst I'm here, I noticed a similar problem with excess properties:

`apps/playground/src/main.ts`:
```ts
api.GET('/users/{username}', {
  params: {
    path: {
      username: 'naoufal',
    },
  },
  // ❌ Missing type error here
  excessProperty: 'foo',
});

```

This is due to `Init & { [key: string]: unknown }`. Can we remove the unknown record part?
